### ISSUE TITLE
New rule: disallow push in Array.forEach

### ISSUE THREAD

**Issue by remcohaszing** [eslint/eslint#7462]
**Please describe what the rule should do:**

The rule should disallow pushing results from `Array.prototype.forEach` to another array. Instead, `Array.prototype.map` should be used.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

``` js
// Disallowed
const result = [];
input.forEach(item => {
  result.push(item + 1);
});

// Alternative construct
const result = input.map(item => item +1);
```

``` js
// Disallowed
const result = [];
input.forEach(item => {
  if (item > 10) {
    result.push(item + 1);
  }
});

// Alternative construct
const result = input
  .filter(item => item > 10)
  .map(item => item +1);
```

**Why should this rule be included in ESLint (instead of a plugin)?**
- It is generic to a core part of JavaScript.
- I've manually detected the disallowed pattern many times in code review. Many people don't seem to consider the alternative.
- It doesn't fit in any existing plugin I'm aware of.

**Comment by ljharb** [comment]
How can you know it's Array's `forEach`? What about `Map` or `Set`, or an arbitrary `forEach` function on any random object?

**Comment by platinumazure** [comment]
[@ljharb's comment](https://github.com/eslint/eslint/issues/7462#issuecomment-256583231) is worth noting here. Due to that risk of false positives, I'm not sure this is a good core rule. (My position: Slight :-1: but open to being convinced.)

You are certainly welcome to create a custom rule if the ESLint team doesn't accept this proposal.

**Comment by ljharb** [comment]
(don't get me wrong, i'd love this rule, but it's not syntactic so there's no way to know you're calling that method on an array)

**Comment by Wilfred** [comment]
FWIW, `array-callback-return` has exactly the same limitation.

**Comment by not-an-aardvark** [comment]
True, but functions with names like `reduce`, `map` and `filter` generally imply that a value must be returned anyway, regardless of whether the caller is an array. With `forEach`, there's no guarantee that `push` should be discouraged.

**Comment by remcohaszing** [comment]
In the end, it's up to the user to decide whether or not to apply this rule to a project.

I've encountered situations like the example code much more often than the use of a map or set, not even talking about looping them. I'd happily add an `eslint-disable-line` comment for those rare occurrences. I think more people would.

**Comment by not-an-aardvark** [comment]
So just to clarify, would this rule disallow any call to a `.push()` function from inside a `.forEach()` function? There are a few other cases where it can't be directly translated to a map/filter:

```js
foo.forEach(element => {
  doSomethingWithSideEffects();
  bar.push(element);
});
```

I'm not sure how I feel about the rule yet; I agree that using something like `.map()` is generally better than `.forEach()` and `.push()`, but I'm concerned that we won't be able to detect this accurately enough.

**Comment by nzakas** [comment]
:-1: I think this is far too specific a use case for a core rule, so I'd suggest creating a custom rule for this instead.

**Comment by alberto** [comment]
Closing since the general consensus is this doesn't belong to core. Thanks for your contribution, anyways.
